pkg/auth/providers/azure/azure_provider.go

```diff
@@ -422,102 +428,103 @@ func (ap *azureProvider) getAzureConfigK8s() (*v32.AzureADConfig, error) {
 	return storedAzureADConfig, nil
 }
 
-func (ap *azureProvider) userToPrincipal(user graphrbac.User) v3.Principal {
-	p := v3.Principal{
-		ObjectMeta:    metav1.ObjectMeta{Name: Name + "_user://" + *user.ObjectID},
-		DisplayName:   *user.DisplayName,
-		LoginName:     *user.UserPrincipalName,
-		PrincipalType: "user",
-		Provider:      Name,
-	}
-	return p
-}
+// userGroupsToPrincipals attempts to convert a value representing a collection of groups to a slice of principal values.
+// It also stores group values in the in-memory package-level cache for faster subsequent access.
+func userGroupsToPrincipals(azureClient clients.AzureClient, groupNames []string) ([]v3.Principal, error) {
+	var tasksManager errgroup.Group
+	groupPrincipals := make([]v3.Principal, len(groupNames))
 
-func (ap *azureProvider) userGroupsToPrincipals(azureClient *azureClient, groups graphrbac.UserGetMemberGroupsResult) ([]v3.Principal, error) {
 	start := time.Now()
 	logrus.Debug("[AZURE_PROVIDER] Started gathering users groups")
 
-	var g errgroup.Group
-	groupPrincipals := make([]v3.Principal, len(*groups.Value))
-	for i, group := range *groups.Value {
+	for i, id := range groupNames {
+		if id == "" {
+			continue
+		}
+
 		j := i
-		gp := group
+		groupID := id
 
-		// Check the cache first, if it exists that saves an API call to azure
-		if principal, ok := groupCache.Get(gp); ok {
-			p := principal.(v3.Principal)
+		if principal, ok := groupCache.Get(groupID); ok {
+			p, ok := principal.(v3.Principal)
+			if !ok {
+				logrus.Errorf("failed to convert a cached group to principal")
+				continue
+			}
 			groupPrincipals[j] = p
 			continue
 		}
 
-		g.Go(func() error {
-			groupObj, err := azureClient.groupClient.Get(context.Background(), gp)
+		tasksManager.Go(func() error {
+			defer func() {
+				if err := recover(); err != nil {
+					logrus.Errorf("[AZURE_PROVIDER] Panic while fetching/caching a group: %v", err)
+				}
+			}()
+			// This is inefficient for a collection of msgraph.Group. This is temporary - until support for Azure AD Graph is removed.
```

Comment:
You might be able to get around this by caching the groups in the client when they are queried.

E.x. 
1. call to azureClient.ListGroupMemberships -> []string, but sets azureClient.Groups = []group
2. Call to azureClient.GetGroup(groupID) checks azureClient.Groups to see if it already listed this group before.

This way, the interface to the provider remains the same, but you still don't have to go all the way to the api. You probably will need to use a map[string]group rather than a slice, but I think that you get the picture.

Reply:
Resolving - we are leaving as is in order to not significantly mangle the old side functionality.